Sources/Build/ManifestBuilder.swift

```diff
@@ -399,26 +403,44 @@ extension LLBuildManifestBuilder {
                                               dependencyArtifactMap: inout SwiftDriver.ExternalDependencyArtifactMap
     ) {
         for dependency in target.dependencies {
-            guard let dependencyTarget = dependency.target else {
-                fatalError("Expected dependency target: \(dependency.description)")
-            }
-            // Only other swift targets will have corresponding dependency maps
-            guard case .swift(let dependencySwiftTargetDescription) =
-                    plan.targetMap[dependencyTarget] else {
-                continue
-            }
-            guard let dependencyGraph = targetDepGraphMap[dependencyTarget] else {
-                fatalError("Expected dependency graph for target: \(dependency.description)")
+            switch dependency {
+                case .product:
+                    // Product dependencies are broken down into the targets that make them up.
+                    guard let dependencyProduct = dependency.product else {
+                        fatalError("Unknown Dependency Kind: \(dependency.description)")
+                    }
+                    for dependencyProductTarget in dependencyProduct.targets {
+                        addTargetDependencyInfo(for: dependencyProductTarget,
+                                                targetDepGraphMap: targetDepGraphMap,
+                                                dependencyArtifactMap: &dependencyArtifactMap)
+
+                    }
+                case .target:
+                    addTargetDependencyInfo(for: target,
+                                            targetDepGraphMap: targetDepGraphMap,
+                                            dependencyArtifactMap: &dependencyArtifactMap)
             }
-            let moduleName = dependencyTarget.name
-            let dependencyModulePath = dependencySwiftTargetDescription.moduleOutputPath
-            dependencyArtifactMap[ModuleDependencyId.swiftPlaceholder(moduleName)] =
-                  (dependencyModulePath, dependencyGraph)
+        }
+    }
 
-            collectTargetDependencyInfos(for: dependencyTarget,
-                                         targetDepGraphMap: targetDepGraphMap,
-                                         dependencyArtifactMap: &dependencyArtifactMap)
+    private func addTargetDependencyInfo(for target: ResolvedTarget,
+                                         targetDepGraphMap: [ResolvedTarget: InterModuleDependencyGraph],
+                                         dependencyArtifactMap: inout SwiftDriver.ExternalDependencyArtifactMap) {
+        guard case .swift(let dependencySwiftTargetDescription) =
+                plan.targetMap[target] else {
+            return
         }
+        guard let dependencyGraph = targetDepGraphMap[target] else {
+            fatalError("Expected dependency graph for target: \(target.description)")
```

Comment:
Hmm. I get nervous about seeing `fatalError` in library code like this. Could incorrect data coming back from `swift-frontend` trigger this `fatalError`, or is this only triggered if there's a bug in SwiftPM's handling of explicit modules itself? The latter is okay for `fatalError`, the former is not.

Reply:
This will occur if SwiftPM got the ordering of build planning targets incorrectly, somehow. That is, if we are trying to plan a build for a target that has dependencies which have not yet been planned (and therefore do not yet have these artifacts). 

If dependencies are processed topologically, correctly, then this structure is populated as we go and the error condition should not arise.

Reply:
Then that's a perfect use of `fatalError`, thanks for clarifying!